### PR TITLE
[recipe,sglang] feat: add Truncated importance sampling + sglang recipe

### PR DESCRIPTION
**Pull request by eternally-z**
### What does this PR do?

This PR introduces a recipe for Truncated Importance Sampling (TIS) rollout using SGLang as the inference engine. 

Previous TIS configurations and experiments primarily used vLLM as the inference engine. For the detailed principles regarding TIS, please refer to PR #2953. We present the first  TIS rollout recipe using SGLang and provide the experimental results below. 

We added a patch in SGLang that applies flash-rl to quantization, supporting various FP8 granularities.Please refer to sglang PR [9650](https://github.com/sgl-project/sglang/pull/9650) & [15440](https://github.com/sgl-project/sglang/pull/15440)(implementation) and  [14870](https://github.com/sgl-project/sglang/pull/14870)(logic abstraction).
#### Configuration
- Model: [Qwen/Qwen3-8B-Base](https://huggingface.co/Qwen/Qwen3-8B-Base)
- Training Recipe: DAPO
- Training Dataset: DAPO-Math-17k
- Quantization Scheme: dynamic blockwise fp8
- Validation: AIME-2024
- Prompt batch size 32, n=16.
- Rollout batch size: 32 \* 3 \* 16
- Train_batch_size & ppo_mini_batch_size 32
- Token-level TIS, C=2
- 8*H20. veRL

#### Experiments results 
@Wilboludriver @AniZpZ

#### Observations
Accuracy of Quantization: Increasing val scores and reasonable response logs confirm (not shown here) that training precision remains intact under Blockwise FP8 Rollout. In contrast, Per-channel FP8 quantization leads to significant precision degradation during the generation phase.

Effects on Training: With the same number of training steps, FP8 Rollout yields longer response lengths and consistently higher AIME-2024 val scores compared to BF16. This performance boost might be attributed to noise introduced by FP8; further investigation using mismatch metrics is required.

Gen. Throughput: FP8 Rollout initially exhibits higher throughput than BF16 but is overtaken in later stages. Profiling is currently underway to identify the bottleneck.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by wuxibin89** [comment]
Better move the script to `examples/rollout_correction` folder?

**Comment by wuxibin89** [comment]
> With TIS the mismatch between per-channel quantized FP8 rollout and native BF16 rollout gets alleviated, which can be also inferred from the rewards curves.

Do you implement sglang fp8 rollout? We already have an implementation in https://github.com/volcengine/verl/pull/4415

**Comment by szrlee** [comment]
@eternally-z please take a look on https://verl.readthedocs.io/en/latest/algo/rollout_corr.html

**Comment by eternally-z** [comment]
> @eternally-z please take a look on https://verl.readthedocs.io/en/latest/algo/rollout_corr.html

Thanks for catching this. The API usage error was due to my local environment running an outdated version of verl. I have updated the code to match the latest API and pushed the fix.

**Comment by AniZpZ** [comment]
> > With TIS the mismatch between per-channel quantized FP8 rollout and native BF16 rollout gets alleviated, which can be also inferred from the rewards curves.
> 
> Do you implement sglang fp8 rollout? We already have an implementation in #4415

we have a [pr](https://github.com/sgl-project/sglang/pull/9650) that adds a patch in sglang that apply flash-rl in sglang quantization which supporting various FP8 quantization granularities, while remaining completely non-intrusive to verl

**Comment by eternally-z** [comment]
> Better move the script to `examples/rollout_correction` folder?

Thanks for the suggestion.The script has been moved to examples/rollout_correction/ .
